src/dataaccess/McodeV05SmartOnFhirDataSource.jsx

```diff
@@ -0,0 +1,124 @@
+        if (this._client) return Promise.resolve(this._client);
+
+        return new Promise((resolve, _reject) => {
+            // this is admittedly clunky, but the FHIR client function uses callbacks
+            // so this converts it to a Promise
+            window.FHIR.oauth2.ready((smart) => {
+                resolve(smart);
+            });
+        }).then((client) => {
+            this._client = client;
+            return this._client;
+        });
+    }
+
+    getPatient(id, callback) {
+        this._getClientAsync().then(client => client.patient.api.conformance({}))
+        .then(metadata => metadata.data.rest[0].resource.map(res => res.type))
+        .then(resourceTypes => {
+            // the FHIR client library does not seem to support calling Patient$everything, so we fake it by manually fetching all the resource types
+            // further, the library only supports async querying
+            const queries = [];
+            for (const resourceType of resourceTypes) {
+                const result = this._client.patient.api.search({ type: resourceType, query: {} });
+                queries.push(result);
+            }
+
+            return Promise.all(queries);
+        }).then(responses => {
+            const entries = [];
+            responses.forEach(response => {
+                // response.data is a Bundle of type searchset
+                // TODO: error handling?
+                if (response && response.data && response.data.entry) {
+                    entries.push(...response.data.entry);
+                }
+            });
+
+            // rather than update PatientRecord and all the Flux wrapper classes to take in FHIR,
+            // the approach here is:
+            // 1. load them as FHIR as regular SHR objects,
+            // 2. convert those regular SHR objects into SHR json
+            // 3. instantiate the PatientRecord with the SHR JSON so that it hydrates the Flux wrapper classes correctly
+            // this approach will likely result in slower performance than supporting FHIR --> Flux wrappers directly, but with massive development time savings
+
+            const mappedResources = {};
+            const referencesOut = [];
+            const allObjects = entries.map(entry => {
+                try {
+                    const result = ObjectFactory.createInstanceFromFHIR(entry.resource, null, this._client.patient.id, entries, mappedResources, referencesOut);
+                    mappedResources[entry.fullUrl] = result;
+                    return result;
+                } catch (e) {
+                    // just log the error, don't stop processing other potentially good objects
+                    console.error(e);
+                    return null;
+                }
+            });
+
+            allObjects.push(...referencesOut);
+            const json = allObjects.filter(o => o).map(o => o.toJSON());
+            return new PatientRecord(json);
```

Comment:
The way this looks to work is to use ObjectFactory to create the mcode objects,  these are then serialized to json, passed off to the PatientRecord which then calls the FluxObjectFactory which also delegates to the ObjectFactory to turn them back into Objects.   Is there a reason for the creation -> serialization -> creation   workflow?

Reply:
It's because the PatientRecord contains Flux wrappers (ex FluxPatient, FluxPerson, etc) not just the plain ES6 model classes, and the wrappers don't contain support to load them from FHIR. It would be nice to be able to support FHIR -> Flux wrappers directly, but that would require modifying all of the wrapper classes and object factories, then everything that calls them, etc.